### PR TITLE
Very first implementation of HttpClient listener

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
For now only supports netstandard2.0. Propagation not integrated yet.

### CODE REVIEW

File: src/OpenCensus.Collector.Dependencies/Implementation/DiagnosticSourceListener.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                {
+                    this.handler.OnStartActivity(Activity.Current, value.Value);
+                }
+                else if (value.Key.EndsWith("Stop"))
+                {
+                    this.handler.OnStopActivity(Activity.Current, value.Value);
+                }
+                else if (value.Key.EndsWith("Exception"))
+                {
+                    this.handler.OnStopActivityWithException(Activity.Current, value.Value);
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/DiagnosticSourceListener.cs]: Exception does not stop Activity

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            span.PutHttpHostAttribute(request.RequestUri.Host);
+            span.PutHttpPathAttribute(request.RequestUri.AbsolutePath);
+            span.PutHttpUrlAttribute(request.RequestUri.ToString());
+        }
+
+        public override void OnStopActivity(Activity activity, object payload)
+        {
+            if (!(this.stopResponseFetcher.Fetch(payload) is HttpResponseMessage response))
+            {
+                // Debug.WriteLine("response is null");
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: response could be null for DNS issues, timeouts, etc...

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            }
+
+            var requestTaskStatus = this.stopRequestStatusFetcher.Fetch(payload) as TaskStatus?;
+
+            var span = this.Tracer.CurrentSpan;
+            if (requestTaskStatus.HasValue)
+            {
+                if (requestTaskStatus != TaskStatus.RanToCompletion)
+                {
+                    span.PutErrorAttribute(requestTaskStatus.ToString());
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: error attribute is recommended to be bool.

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                }
+            }
+
+            if ((int)response.StatusCode >= 200 && (int)response.StatusCode <= 301)
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: i'd suggest adding helper method to convert http status code to OC status - it will be useful in other places: Asp.NET and Asp.Net Core

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            else if ((int)response.StatusCode == 501)
+            {
+                span.Status = Status.Unimplemented;
+            }
+            else
+            {
+                span.Status = Status.Unknown;
+            }
+
+            span.Status = span.Status.WithDescription(response.StatusCode + " " + response.ReasonPhrase);
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: assuming task has faulted and there is no response, we should add exception details here (in onException handler)

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                return;
+            }
+
+            var span = this.Tracer.CurrentSpan;
+            var requestTaskStatus = this.stopRequestStatusFetcher.Fetch(payload) as TaskStatus?;
+            if (requestTaskStatus.HasValue)
+            {
+                if (requestTaskStatus != TaskStatus.RanToCompletion)
+                {
+                    span.PutErrorAttribute(requestTaskStatus.ToString());
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: bool

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            }
+
+            this.LocalScope.Value?.Dispose();
+
+            if (exc is HttpRequestException)
+            {
+                // this will be System.Net.Http.WinHttpException: The server name or address could not be resolved
+                // on netstandard...
+                if (exc.InnerException is WebException &&
+                    ((WebException)exc.InnerException).Status == WebExceptionStatus.NameResolutionFailure)
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: would it work on Core? It did not work 1,5 years ago. Also it'd be great to have WebException status in the span status.

Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: nope. I put a comment for this

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 6 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: not following.

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 5 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: definitely

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 4 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: yep

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 3 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: do you have a link to spec?

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 7 above)
Comment (lmolkova) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: so perhaps this would never happen on this DiagnosticListener - this is core only

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 3 above)
Comment (lmolkova) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        public override void OnStopActivity(Activity activity, object payload)
+        {
+            var span = this.Tracer.CurrentSpan;
+
+            if (!(this.stopResponseFetcher.Fetch(payload) is HttpResponseMessage response))
+            {
+                // response could be null for DNS issues, timeouts, etc...
+                // TODO: how do we make sure we will not close a scope that wasn't opened?
+
+                span.End();
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: span status for faulted/cacelled tasks is set after this - i.e. will never be executed

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            {
+                // TODO: on netstandard this will be System.Net.Http.WinHttpException: The server name or address could not be resolved
+                if (exc.InnerException is WebException &&
+                    ((WebException)exc.InnerException).Status == WebExceptionStatus.NameResolutionFailure)
+                {
+                    span.Status = Status.InvalidArgument;
+                }
+                else if (exc.InnerException != null)
+                {
+                    span.Status = Status.Unknown.WithDescription(exc.Message);
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: this probably conflicts with status assignment in OnStop. We should do it just here or just there

Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: it's OK to override it from the response code. This code runs for DNS error for instance

Comment (lmolkova) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: oh, i see you override just the status, but not error message

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+        public override void OnException(Activity activity, object payload)
+        {
+            if (!(this.stopExceptionFetcher.Fetch(payload) is Exception exc))
+            {
+                // Debug.WriteLine("response is null");
+                return;
+            }
+
+            var span = this.Tracer.CurrentSpan;
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: please check for null. even though it should never happen, but there may be some bug that will cause current span to be null.

Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: done

File: src/OpenCensus.Collector.Dependencies/Implementation/ListenerHandler.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        public string SourceName { get; }
+
+        public virtual void OnStartActivity(Activity activity, object payload)
+        {
+            this.Tracer.SpanBuilder(activity.OperationName).SetSampler(this.Sampler).StartScopedSpan();
+        }
+
+        public virtual void OnStopActivity(Activity activity, object payload)
+        {
+            var span = this.Tracer.CurrentSpan;
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/ListenerHandler.cs]: span might be null (should not, but we've seen it before)

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                return;
+            }
+
+            // TODO: this needs to be generalized
+            if (request.RequestUri.ToString().Contains("zipkin.azurewebsites.net"))
+            {
+                return;
+            }
+
+            this.Tracer.SpanBuilder(request.RequestUri.AbsolutePath).SetSampler(this.Sampler).StartScopedSpan();
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: this is not part of this PR, of course, but I wonder if we can change the API `his.Tracer.SpanBuilder(request.RequestUri.AbsolutePath).SetSampler(this.Sampler).StartScopedSpan();` to return the Span explicitly.

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        private readonly PropertyFetcher stopExceptionFetcher = new PropertyFetcher("Exception");
+        private readonly PropertyFetcher stopRequestStatusFetcher = new PropertyFetcher("RequestTaskStatus");
+
+        public HttpHandlerDiagnosticListener(ITracer tracer, ISampler sampler) : base("HttpHandlerDiagnosticListener", tracer, sampler)
+        {
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
```
Comment (lmolkova) [inline src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: the cost of having a bug here (in any callback) is extremely high. So if we accidentally throw, it will break user app and prevent the http call. We should put all callbacks in try/catch and log exceptions.

File: src/OpenCensus.Collector.Dependencies/Implementation/ListenerHandler.cs (hunk 11 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/ListenerHandler.cs]: yep. it's handled

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 12 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: So basically to make `Span` disposable? I'd rather have optional out parameter there so fxcop will no run wild with many undisposed objects detection. Would you agree?

Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: https://github.com/census-instrumentation/opencensus-csharp/issues/45

File: src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs (hunk 13 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/Implementation/HttpHandlerDiagnosticListener.cs]: It is handled in `DiagnosticSourceListener`

File: src/OpenCensus.Collector.Dependencies/Implementation/ListenerHandler.cs (hunk 11 above)
Comment (lmolkova) [reply src/OpenCensus.Collector.Dependencies/Implementation/ListenerHandler.cs]: where? I believe it's handled in http case, but not in the generic case
